File: src/main/java/ch/softappeal/yass/transport/socket/SocketTransport.java
package ch.softappeal.yass.transport.socket;

import ch.softappeal.yass.serialize.Reader;
import ch.softappeal.yass.serialize.Serializer;
import ch.softappeal.yass.serialize.Writer;
import ch.softappeal.yass.transport.PathResolver;
import ch.softappeal.yass.transport.PathSerializer;
import ch.softappeal.yass.transport.TransportSetup;
import ch.softappeal.yass.util.Check;
import ch.softappeal.yass.util.Exceptions;

import java.io.OutputStream;
import java.net.Socket;
import java.util.concurrent.Executor;

/**
 * Each session gets its own socket.
 */
public final class SocketTransport extends SocketListener {

    private final SocketConnection.Factory connectionFactory;
    private final Serializer pathSerializer;
    private final PathResolver pathResolver;

    /**
     * @param readerExecutor used once for each session
     */
    public SocketTransport(
        final Executor readerExecutor, final SocketConnection.Factory connectionFactory,
        final Serializer pathSerializer, final PathResolver pathResolver
    ) {
        super(readerExecutor);
        this.connectionFactory = Check.notNull(connectionFactory);
        this.pathSerializer = Check.notNull(pathSerializer);
        this.pathResolver = Check.notNull(pathResolver);
    }

    public SocketTransport(
        final Executor readerExecutor, final SocketConnection.Factory connectionFactory,
        final TransportSetup setup
    ) {
        this(
            readerExecutor, connectionFactory,
            PathSerializer.INSTANCE, new PathResolver(PathSerializer.DEFAULT, setup)
        );
    }

    @Override void accept(final Socket socket) throws Exception {
        final Reader reader = Reader.create(socket.getInputStream());
        final TransportSetup setup = pathResolver.resolvePath(pathSerializer.read(reader));
        SocketConnection.create(connectionFactory, setup, socket, reader, socket.getOutputStream());
    }

    public static void connect(
        final Executor readerExecutor, final SocketConnection.Factory connectionFactory, final TransportSetup setup, final SocketConnector socketConnector,
        final Serializer pathSerializer, final Object path
    ) {
        Check.notNull(connectionFactory);
        Check.notNull(setup);
        Check.notNull(pathSerializer);
        Check.notNull(path);
        try {
            SocketUtils.execute(readerExecutor, socketConnector.connect(), socket -> {
                final OutputStream out = socket.getOutputStream();
                pathSerializer.write(path, Writer.create(out));
                out.flush();
                SocketConnection.create(connectionFactory, setup, socket, Reader.create(socket.getInputStream()), out);
            });
        } catch (final Exception e) {
            throw Exceptions.wrap(e);
        }
    }

    public static void connect(final Executor readerExecutor, final SocketConnection.Factory connectionFactory, final TransportSetup setup, final SocketConnector socketConnector) {
        connect(
            readerExecutor, connectionFactory, setup, socketConnector,
            PathSerializer.INSTANCE, PathSerializer.DEFAULT
        );
    }

}

File: kotlin/yass/main/ch/softappeal/yass/transport/socket/SocketTransport.kt
package ch.softappeal.yass.transport.socket

import ch.softappeal.yass.*
import ch.softappeal.yass.remote.*
import ch.softappeal.yass.serialize.*
import ch.softappeal.yass.transport.*
import java.io.*
import java.net.*
import java.util.concurrent.*

private val socket_ = ThreadLocal<Socket>()

val socket: Socket get() = checkNotNull(socket_.get()) { "no active invocation" }

/** Buffering of output is needed to prevent long delays due to Nagle's algorithm. */
private fun createBuffer(): ByteArrayOutputStream = ByteArrayOutputStream(128)

private fun write(buffer: ByteArrayOutputStream, socket: Socket) {
    val out = socket.getOutputStream()
    buffer.writeTo(out)
    out.flush()
}

/** [requestExecutor] is called once for each request. */
fun socketServer(setup: ServerSetup, requestExecutor: Executor) = object : SocketListener(requestExecutor) {
    override fun accept(socket: Socket) = threadLocal(socket_, socket) {
        val reader = reader(socket.getInputStream())
        val transport = setup.resolve(reader)
        transport.invocation(true, transport.read(reader)).invoke({ socket.close() }) { reply ->
            val buffer = createBuffer()
            transport.write(writer(buffer), reply)
            write(buffer, socket)
        }
    }
}

fun socketClient(setup: ClientSetup, socketConnector: SocketConnector) = object : Client() {
    override fun executeInContext(action: () -> Any?): Any? {
        val socket = socketConnector()
        try {
            setForceImmediateSend(socket)
            return threadLocal(socket_, socket) { action() }
        } catch (e: Exception) {
            close(socket, e)
            throw e
        }
    }

    /**
     * If an exception is throw in this method, the socket would also be closed in [executeInContext].
     * That's ok because [Socket.close] is idempotent.
     */
    override fun invoke(invocation: ClientInvocation) = socket_.get().use { socket ->
        invocation.invoke(true) { request ->
            val buffer = createBuffer()
            setup.write(writer(buffer), request)
            write(buffer, socket)
            invocation.settle(setup.read(reader(socket.getInputStream())))
        }
    }
}
